### PR TITLE
Feat+version

### PR DESCRIPTION
**Pull request by nojhan**
Implements #749 by providing `_lp_version_greatereq` and `_lp_version_string`.

### CODE REVIEW

**Comment by nojhan** [comment]
Pushed for sharing, but so far the tests fail under Zsh, and I'm not sure to understand why.

File: liquidprompt [hunk 1]
```diff
@@ -154,6 +156,35 @@ esac
 # CONFIGURATION #
 #################
 
+# Return true if the current Liquid Prompt version is greater than or equal to the specified version.
+# [OPTIONAL?] Return 1 if minor or less version difference, 2 if major difference.
+_lp_version_greatereq() {  # major, minor, [patch], [string], [number]
+    local major="$1" minor="$2" patch="${3-}" string="${4-}" number="${5-}"
+    if (( major > _LP_VERSION[0] )); then
```
Comment (Rycieos) [inline liquidprompt:163]: I forgot when writing my PoC, but this needs to use the `_LP_FIRST_INDEX` or whatever it is variable. That is probably why Zsh is failing tests.

File: liquidprompt [hunk 2]
```diff
@@ -992,6 +1023,25 @@ _lp_grep_fields() {
...
+        local number="${_LP_VERSION[_LP_FIRST_INDEX+4]-}"
+    else
+        local major="${1}"
+        local minor="${2}"
+        local patch="${3-}"
+        local string="${4-}"
+        local number="${5-}"
+    fi
+    #shellcheck disable=SC2034
+    lp_version="${major}.${minor}.${patch}-${string}-${number}"
```
Comment (Rycieos) [inline liquidprompt:1042]: This needs more logic, the string and number part will not always be set.

File: tests/test_utils.sh [hunk 3]
```diff
@@ -783,6 +783,41 @@ function test_grep_fields {
...
+    _lp_version_greatereq 1 1 3 beta 4
+    assertTrue "greater version minor" "$?"
+    _lp_version_greatereq 0 2 3 beta 4
+    assertTrue "greater version major" "$?"
+
+    _lp_version_string # Defaults to _LP_VERSION
+    assertEquals "1.2.3-beta-4" "$lp_version"
+
+    _lp_version_string 5 4 3 rc 2
+    assertEquals "5.4.3-rc-2" "$lp_version"
```
Comment (Rycieos) [inline tests/test_utils.sh:818]: Needs another test for the case of `_lp_version_string 5 4 3` or similar.

**Review by Rycieos — CHANGES_REQUESTED** [review]

File: docs/functions/util.rst [hunk 4]
```diff
@@ -143,40 +180,27 @@ These functions are designed to be used by themes.
...
+   .. versionadded:: 2.2
 
-   Colorize the given string with a color depending on its hash. The color is
-   chosen among: (green, yellow, blue, purple, cyan). Note that the red color is
-   not a candidate, as it should be reserved for alerts.
+.. function:: _lp_version_string([major, [minor, [patch, [string, [number]]]]]) -> var:lp_version
 
-   .. versionadded:: 2.2
+   Formats the given version number in a version string of the form:
+   "${major}.${minor}.${patch}-${string}-${number}"
```
Comment (Rycieos) [inline docs/functions/util.rst:198]: Semantic versioning allows for either, but I have been formatting the version with a period, as `${string}.${number}`, not with a hyphen. So `alpha.1`. Might as well be consistent.

File: liquidprompt [hunk 5]
```diff
@@ -1057,6 +1088,34 @@ _lp_fill() {
...
+        local number="${5-}"
+    fi
+    #shellcheck disable=SC2034
+    lp_version="${major}.${minor}"
+    if [[ -n "$patch" ]]; then
+        lp_version+=".${patch}"
+        if [[ -n "$string" ]]; then
+            lp_version+="-${string}"
+            if [[ -n "$number" ]]; then
+                lp_version+="-${number}"
```
Comment (Rycieos) [inline liquidprompt:1113]: Same as above:
```suggestion
                lp_version+=".${number}"
```

File: docs/functions/util.rst [hunk 6]
```diff
@@ -143,40 +180,27 @@ These functions are designed to be used by themes.
...
-
-   .. versionadded:: 2.0
-
-.. function:: _lp_raw_title(title)
+.. function:: _lp_version_greatereq([major, [minor, [patch, [string, [number]]]]])
 
-   Sets the theme generated title to *title*. The input is not escaped in any
-   way: if the input contains terminal formatting, use
-   :func:`_lp_formatted_title` instead.
+   Returns true (0) if Liquid Prompt version is greater than or equal to the the given version.
```
Comment (Rycieos) [inline docs/functions/util.rst:185]: Line too long.

**Review by Rycieos — CHANGES_REQUESTED** [review]
Looks good, just needs a rebase to fix the spelling issue.

File: docs/functions/util.rst (hunk 4 above)
Comment (nojhan) [reply docs/functions/util.rst:198]: Should we keep the hyphen between patch and string?

Comment (Rycieos) [reply docs/functions/util.rst:198]: Yes, that is required by Semantic versioning, and what I have been doing. See tag [v2.1.0-rc.1](https://github.com/nojhan/liquidprompt/tree/v2.1.0-rc.1) for example.

File: docs/functions/util.rst [hunk 7]
```diff
@@ -143,40 +180,28 @@ These functions are designed to be used by themes.
...
-   :func:`__lp_strip_escapes` to strip terminal formatting from being added to
-   the title.
-
-   This function will do nothing and return ``2`` if :attr:`LP_ENABLE_TITLE`
-   is disabled.
-
-   .. versionadded:: 2.0
-
-.. function:: _lp_raw_title(title)
+.. function:: _lp_version_greatereq([major, [minor, [patch, [string, [number]]]]])
```
Comment (Rycieos) [inline docs/functions/util.rst:183]: This signature does not match the inline docs and code. Major and minor are not optional.

```suggestion
.. function:: _lp_version_string(major, minor, [patch, [string, [number]]]) -> var:lp_version
```

File: docs/functions/util.rst [hunk 8]
```diff
@@ -143,40 +180,28 @@ These functions are designed to be used by themes.
...
+   or equal to the the given version.
+   Returns 1 (false) if there is a *minor* or less version difference,
+   and 2 (false) if it is a *major* difference.
 
-   This function will do nothing and return ``2`` if :attr:`LP_ENABLE_TITLE`
-   is disabled.
+   See also :func:`_lp_version_string`.
 
-   .. versionadded:: 2.0
+   .. warning:: This only supports the following strings: "alpha", "beta" and "rc".
```
Comment (Rycieos) [inline docs/functions/util.rst:192]: Clearer meaning
```suggestion
   .. warning:: This only supports the following input values for `string`: "alpha", "beta" and "rc".
```

**Review by Rycieos — CHANGES_REQUESTED** [review]
Looks good, just a minor docs issue.

File: docs/functions/util.rst (hunk 7 above)
Comment (nojhan) [reply docs/functions/util.rst:183]: Your diff is ambiguous. I corrected `_lp_version_greatereq`, but for `_lp_version_string`, all parametert are optional (to allow for collecting the current version string).

**Review by Rycieos — APPROVED** [review]

**Comment by Rycieos** [comment]
Just needs squashing I think.

**Review by Rycieos — APPROVED** [review]

**Pull request merged**